README
NAME
    Toolforge::MixNMatch::Struct - Toolforge Mix'n'match tool structures.

DESCRIPTION
    This distributions is set of basic conversions from/to Toolforge
    Mix'n'match tool JSON structures. (https://mix-n-match.toolforge.org/).

SEE ALSO
    Toolforge::MixNMatch::Struct::Catalog
        Mix'n'match catalog structure serialization.

    Toolforge::MixNMatch::Struct::User
        Mix'n'match user structure serialization.

    Toolforge::MixNMatch::Struct::YearMonth
        Mix'n'match year/month structure serialization.

REPOSITORY
    <https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © Michal Josef Špaček 2020

    BSD 2-Clause License

VERSION
    0.05



Changes
0.05
 - Update META.yml file.
 - Update Module::Install to 1.21 version.

0.04 2020-12-22T12:11:00+01:00
 - Add checks for object in obj2struct() in all classes.
 - Add doc for Catalog. And examples for it.
 - Add doc for User. And examples for it.
 - Add doc for YearMonth. And examples for it.
 - Export obj2struct() and struct2obj().

0.03 2020-12-20T12:00:07+01:00
 - Fix dependencies.

0.02 2020-12-19T18:17:52+01:00
 - Add tests for Catalog.
 - Add tests for User.
 - Add tests for YearMonth.
 - Fix conversion of month in YearMonth class.

0.01 2020-12-19T17:20:09+01:00
 - First version.


Struct.pm
package Toolforge::MixNMatch::Struct;

use strict;
use warnings;

our $VERSION = 0.05;

1;

__END__

=pod

=encoding utf8

=head1 NAME

Toolforge::MixNMatch::Struct - Toolforge Mix'n'match tool structures.

=head1 DESCRIPTION

This distributions is set of basic conversions from/to Toolforge Mix'n'match tool JSON structures.
(https://mix-n-match.toolforge.org/).

=head1 SEE ALSO

=over

=item L<Toolforge::MixNMatch::Struct::Catalog>

Mix'n'match catalog structure serialization.

=item L<Toolforge::MixNMatch::Struct::User>

Mix'n'match user structure serialization.

=item L<Toolforge::MixNMatch::Struct::YearMonth>

Mix'n'match year/month structure serialization.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© Michal Josef Špaček 2020

BSD 2-Clause License

=head1 VERSION

0.05

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Toolforge Mix\'n\'match tool structures.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Toolforge-MixNMatch-Struct';
perl_version 5.006002;
readme_from 'Struct.pm';
recursive_author_tests('xt');
requires 'Error::Pure' => 0;
requires 'Exporter' => 0;
requires 'Readonly' => 0;
requires 'Toolforge::MixNMatch::Object' => 0.02;
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct/issues',
resources 'homepage' => 'https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct';
resources 'repository' => 'git://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.05';

# Run.
WriteAll();


Struct/User.pm
package Toolforge::MixNMatch::Struct::User;

use base qw(Exporter);
use strict;
use warnings;

use Error::Pure qw(err);
use Readonly;
use Toolforge::MixNMatch::Object::User;

Readonly::Array our @EXPORT_OK => qw(obj2struct struct2obj);

our $VERSION = 0.05;

sub obj2struct {
	my $obj = shift;

	if (! defined $obj) {
		err "Object doesn't exist.";
	}
	if (! $obj->isa('Toolforge::MixNMatch::Object::User')) {
		err "Object isn't 'Toolforge::MixNMatch::Object::User'.";
	}

	my $struct_hr = {
		'cnt' => $obj->count,
		'uid' => $obj->uid,
		'username' => $obj->username,
	};

	return $struct_hr;
}

sub struct2obj {
	my $struct_hr = shift;

	my $obj = Toolforge::MixNMatch::Object::User->new(
		'count' => $struct_hr->{'cnt'},
		'uid' => $struct_hr->{'uid'},
		'username' => $struct_hr->{'username'},
	);

	return $obj;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Toolforge::MixNMatch::Struct::User - Mix'n'match user structure serialization.

=head1 SYNOPSIS

 use Toolforge::MixNMatch::Struct::User qw(obj2struct struct2obj);

 my $struct_hr = obj2struct($obj);
 my $obj = struct2obj($struct_hr);

=head1 DESCRIPTION

This conversion is between object defined in Toolforge::MixNMatch::Object::User and structure
serialized via JSON to Mix'n'match application.

=head1 SUBROUTINES

=head2 C<obj2struct>

 my $struct_hr = obj2struct($obj);

Convert Toolforge::MixNMatch::Object::User instance to structure.

Returns reference to hash with structure.

=head2 C<struct2obj>

 my $obj = struct2obj($struct_hr);

Convert structure of time to object.

Returns Toolforge::MixNMatch::Object::User instance.

=head1 ERRORS

 obj2struct():
         Object doesn't exist.
         Object isn't 'Toolforge::MixNMatch::Object::User'.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use Data::Printer;
 use Toolforge::MixNMatch::Object::User;
 use Toolforge::MixNMatch::Struct::User qw(obj2struct);

 # Object.
 my $obj = Toolforge::MixNMatch::Object::User->new(
         'count' => 6,
         'uid' => 1,
         'username' => 'Skim',
 );

 # Get structure.
 my $struct_hr = obj2struct($obj);

 # Dump to output.
 p $struct_hr;

 # Output:
 # \ {
 #     cnt        6,
 #     uid        1,
 #     username   "Skim"
 # }

=head1 EXAMPLE2

 use strict;
 use warnings;

 use Toolforge::MixNMatch::Struct::User qw(struct2obj);

 # Time structure.
 my $struct_hr = {
        'cnt' => 6,
        'uid' => 1,
        'username' => 'Skim',
 };

 # Get object.
 my $obj = struct2obj($struct_hr);

 # Get count.
 my $count = $obj->count;

 # Get user UID.
 my $uid = $obj->uid;

 # Get user name.
 my $username = $obj->username;

 # Print out.
 print "Count: $count\n";
 print "User UID: $uid\n";
 print "User name: $username\n";

 # Output:
 # Count: 6
 # User UID: 1
 # User name: Skim

=head1 DEPENDENCIES

L<Error::Pure>,
L<Exporter>,
L<Readonly>,
L<Toolforge::MixNMatch::Struct::User>.

=head1 SEE ALSO

=over

=item L<Toolforge::MixNMatch::Struct>

Toolforge Mix'n'match tool structures.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© Michal Josef Špaček 2020

BSD 2-Clause License

=head1 VERSION

0.05

=cut


Struct/YearMonth.pm
package Toolforge::MixNMatch::Struct::YearMonth;

use base qw(Exporter);
use strict;
use warnings;

use Error::Pure qw(err);
use Readonly;
use Toolforge::MixNMatch::Object::YearMonth;

Readonly::Array our @EXPORT_OK => qw(obj2struct struct2obj);

our $VERSION = 0.05;

sub obj2struct {
	my $obj = shift;

	if (! defined $obj) {
		err "Object doesn't exist.";
	}
	if (! $obj->isa('Toolforge::MixNMatch::Object::YearMonth')) {
		err "Object isn't 'Toolforge::MixNMatch::Object::YearMonth'.";
	}

	my $struct_hr = {
		'cnt' => $obj->count,
		'ym' => $obj->year.(sprintf '%02d', $obj->month),
	};

	return $struct_hr;
}

sub struct2obj {
	my $struct_hr = shift;

	my ($year, $month) = $struct_hr->{'ym'} =~ m/^(\d{4})(\d{2})$/ms;
	my $obj = Toolforge::MixNMatch::Object::YearMonth->new(
		'count' => $struct_hr->{'cnt'},
		'month' => int($month),
		'year' => $year,
	);

	return $obj;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Toolforge::MixNMatch::Struct::YearMonth - Mix'n'match year/month structure serialization.

=head1 SYNOPSIS

 use Toolforge::MixNMatch::Struct::YearMonth qw(obj2struct struct2obj);

 my $struct_hr = obj2struct($obj);
 my $obj = struct2obj($struct_hr);

=head1 DESCRIPTION

This conversion is between object defined in Toolforge::MixNMatch::Object::YearMonth and structure
serialized via JSON to Mix'n'match application.

=head1 SUBROUTINES

=head2 C<obj2struct>

 my $struct_hr = obj2struct($obj);

Convert Toolforge::MixNMatch::Object::YearMonth instance to structure.

Returns reference to hash with structure.

=head2 C<struct2obj>

 my $obj = struct2obj($struct_hr);

Convert structure of time to object.

Returns Toolforge::MixNMatch::Object::YearMonth instance.

=head1 ERRORS

 obj2struct():
         Object doesn't exist.
         Object isn't 'Toolforge::MixNMatch::Object::YearMonth'.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use Data::Printer;
 use Toolforge::MixNMatch::Object::YearMonth;
 use Toolforge::MixNMatch::Struct::YearMonth qw(obj2struct);

 # Object.
 my $obj = Toolforge::MixNMatch::Object::YearMonth->new(
         'count' => 6,
         'month' => 9,
         'year' => 2020,
 );

 # Get structure.
 my $struct_hr = obj2struct($obj);

 # Dump to output.
 p $struct_hr;

 # Output:
 # \ {
 #     cnt   6,
 #     ym    202009
 # }

=head1 EXAMPLE2

 use strict;
 use warnings;

 use Toolforge::MixNMatch::Struct::YearMonth qw(struct2obj);

 # Time structure.
 my $struct_hr = {
        'cnt' => 6,
        'ym' => 202009,
 };

 # Get object.
 my $obj = struct2obj($struct_hr);

 # Get count.
 my $count = $obj->count;

 # Get month.
 my $month = $obj->month;

 # Get year.
 my $year = $obj->year;

 # Print out.
 print "Count: $count\n";
 print "Month: $month\n";
 print "Year: $year\n";

 # Output:
 # Count: 6
 # Month: 9
 # Year: 2020

=head1 DEPENDENCIES

L<Error::Pure>,
L<Exporter>,
L<Readonly>,
L<Toolforge::MixNMatch::Struct::YearMonth>.

=head1 SEE ALSO

=over

=item L<Toolforge::MixNMatch::Struct>

Toolforge Mix'n'match tool structures.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© Michal Josef Špaček 2020

BSD 2-Clause License

=head1 VERSION

0.05

=cut


Struct/Catalog.pm
package Toolforge::MixNMatch::Struct::Catalog;

use base qw(Exporter);
use strict;
use warnings;

use Error::Pure qw(err);
use Readonly;
use Toolforge::MixNMatch::Object::Catalog;
use Toolforge::MixNMatch::Struct::User;
use Toolforge::MixNMatch::Struct::YearMonth;

Readonly::Array our @EXPORT_OK => qw(obj2struct struct2obj);

our $VERSION = 0.05;

sub obj2struct {
	my $obj = shift;

	if (! defined $obj) {
		err "Object doesn't exist.";
	}
	if (! $obj->isa('Toolforge::MixNMatch::Object::Catalog')) {
		err "Object isn't 'Toolforge::MixNMatch::Object::Catalog'.";
	}

	my $struct_hr = {
		'type' => [{
			'type' => $obj->type,
			'cnt' => $obj->count,
		}],
		'user' => [],
		'ym' => [],
	};
	foreach my $year_month (@{$obj->year_months}) {
		push @{$struct_hr->{'ym'}},
			Toolforge::MixNMatch::Struct::YearMonth::obj2struct($year_month);
	}
	foreach my $user (@{$obj->users}) {
		push @{$struct_hr->{'user'}},
			Toolforge::MixNMatch::Struct::User::obj2struct($user);
	}

	return $struct_hr;
}

sub struct2obj {
	my $struct_hr = shift;

	my $year_months = [];
	foreach my $year_month_hr (@{$struct_hr->{'ym'}}) {
		push @{$year_months}, Toolforge::MixNMatch::Struct::YearMonth::struct2obj($year_month_hr);
	}
	my $users = [];
	foreach my $user_hr (@{$struct_hr->{'user'}}) {
		push @{$users}, Toolforge::MixNMatch::Struct::User::struct2obj($user_hr);
	}
	my $obj = Toolforge::MixNMatch::Object::Catalog->new(
		'count' => $struct_hr->{'type'}->[0]->{'cnt'},
		'type' => $struct_hr->{'type'}->[0]->{'type'},
		'users' => $users,
		'year_months' => $year_months,
	);

	return $obj;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Toolforge::MixNMatch::Struct::Catalog - Mix'n'match catalog structure serialization.

=head1 SYNOPSIS

 use Toolforge::MixNMatch::Struct::Catalog qw(obj2struct struct2obj);

 my $struct_hr = obj2struct($obj);
 my $obj = struct2obj($struct_hr);

=head1 DESCRIPTION

This conversion is between object defined in Toolforge::MixNMatch::Object::Catalog and structure
serialized via JSON to Mix'n'match application.

=head1 SUBROUTINES

=head2 C<obj2struct>

 my $struct_hr = obj2struct($obj);

Convert Toolforge::MixNMatch::Object::Catalog instance to structure.

Returns reference to hash with structure.

=head2 C<struct2obj>

 my $obj = struct2obj($struct_hr);

Convert structure of time to object.

Returns Toolforge::MixNMatch::Object::Catalog instance.

=head1 ERRORS

 obj2struct():
         Object doesn't exist.
         Object isn't 'Toolforge::MixNMatch::Object::Catalog'.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use Data::Printer;
 use Toolforge::MixNMatch::Object::Catalog;
 use Toolforge::MixNMatch::Struct::Catalog qw(obj2struct);

 # Object.
 my $obj = Toolforge::MixNMatch::Object::Catalog->new(
         'count' => 10,
         'type' => 'Q5',
         'users' => [
                 Toolforge::MixNMatch::Object::User->new(
                         'count' => 6,
                         'uid' => 1,
                         'username' => 'Skim',
                 ),
                 Toolforge::MixNMatch::Object::User->new(
                         'count' => 4,
                         'uid' => 2,
                         'username' => 'Foo',
                 ),
         ],
         'year_months' => [
                 Toolforge::MixNMatch::Object::YearMonth->new(
                         'count' => 2,
                         'month' => 9,
                         'year' => 2020,
                 ),
                 Toolforge::MixNMatch::Object::YearMonth->new(
                         'count' => 8,
                         'month' => 10,
                         'year' => 2020,
                 ),
         ],
 );

 # Get structure.
 my $struct_hr = obj2struct($obj);

 # Dump to output.
 p $struct_hr;

 # Output:
 # \ {
 #     type   [
 #         [0] {
 #             cnt    10,
 #             type   "Q5"
 #         }
 #     ],
 #     user   [
 #         [0] {
 #             cnt        6,
 #             uid        1,
 #             username   "Skim"
 #         },
 #         [1] {
 #             cnt        4,
 #             uid        2,
 #             username   "Foo"
 #         }
 #     ],
 #     ym     [
 #         [0] {
 #             cnt   2,
 #             ym    202009
 #         },
 #         [1] {
 #             cnt   8,
 #             ym    202010
 #         }
 #     ]
 # }

=head1 EXAMPLE2

 use strict;
 use warnings;

 use Toolforge::MixNMatch::Struct::Catalog qw(struct2obj);

 # Time structure.
 my $struct_hr = {
         'user' => [{
                 'cnt' => 6,
                 'uid' => 1,
                 'username' => 'Skim',
         }, {
                 'cnt' => 4,
                 'uid' => 2,
                 'username' => 'Foo',
         }],
         'type' => [{
                 'cnt' => 10,
                 'type' => 'Q5',
         }],
         'ym' => [{
                 'cnt' => 2,
                 'ym' => 202009,
         }, {
                 'cnt' => 8,
                 'ym' => 202010,
         }],
 };

 # Get object.
 my $obj = struct2obj($struct_hr);

 # Get count.
 my $count = $obj->count;

 # Get type.
 my $type = $obj->type;

 # Get user statistics.
 my $users_ar = $obj->users;

 # Get year/month statistics.
 my $year_months_ar = $obj->year_months;

 # Print out.
 print "Count: $count\n";
 print "Type: $type\n";
 print "Count of users: ".(scalar @{$users_ar})."\n";
 print "Count of year/months: ".(scalar @{$year_months_ar})."\n";

 # Output:
 # Count: 10
 # Type: Q5
 # Count of users: 2
 # Count of year/months: 2

=head1 DEPENDENCIES

L<Error::Pure>,
L<Exporter>,
L<Readonly>,
L<Toolforge::MixNMatch::Object::Catalog>,
L<Toolforge::MixNMatch::Struct::User>,
L<Toolforge::MixNMatch::Struct::YearMonth>.

=head1 SEE ALSO

=over

=item L<Toolforge::MixNMatch::Struct>

Toolforge Mix'n'match tool structures.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Toolforge-MixNMatch-Struct>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© Michal Josef Špaček 2020

BSD 2-Clause License

=head1 VERSION

0.05

=cut
